File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex19BulkPartialSuccess.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Dev;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-19 — Partial success and item-level errors.
/// </summary>
/// <remarks>
/// A bulk request where some EINs miss is HTTP 200 with the misses in the envelope's
/// <c>errors</c> — a success that a caller checking only the status code would misread.
/// </remarks>
public sealed class Ex19BulkPartialSuccess : IExample
{
    /// <inheritdoc />
    public string Id => "ex-19";

    /// <inheritdoc />
    public string Title => "Partial success: HTTP 200 carrying item-level failures.";

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        using var context = ExampleContext.WithFixtures();

        var requested = new[]
        {
            Fixtures.Eins.PublicCharity,
            Fixtures.Eins.NoRecord,
            Fixtures.Eins.PublicCharitySecond,
        };

        var result = await context.Client.Nonprofits.CheckBulkAsync(requested);

        Output.Heading("A successful response with failures inside it");
        Output.Field("HTTP status", result.Status);
        Output.Field("threw", "no — this is a success");
        Output.Field("requested", requested.Length);
        Output.Field("matched", result.Organizations.Count);
        Output.Field("not found", result.NotFoundEins.Count);

        Output.Heading("The item-level errors");

        foreach (var detail in result.Errors)
        {
            Output.Field("resource", detail.Resource);
            Output.Field("reason", detail.Reason);
            Output.Field("code", detail.Code);
            Output.Field("eins", string.Join(", ", detail.Eins));

            // The entry exactly as the API sent it, including anything not typed above.
            Output.Field("raw keys", string.Join(", ", detail.ToDictionary().Keys));
        }

        Output.Heading("Accounting for every input");

        var byEin = result.ByEin();

        foreach (var ein in requested)
        {
            Output.Field(ein, byEin.ContainsKey(ein)
                ? "matched"
                : result.NotFoundEins.Contains(ein, StringComparer.Ordinal)
                    ? "reported not found"
                    : "UNACCOUNTED FOR — investigate");
        }

        // Every input should land in exactly one bucket. An input in neither means the
        // API returned something this code does not understand, and that is worth
        // failing loudly over rather than quietly dropping.
        var unaccounted = requested
            .Where(ein => !byEin.ContainsKey(ein) && !result.NotFoundEins.Contains(ein, StringComparer.Ordinal))
            .ToList();

        Output.Field("unaccounted", unaccounted.Count == 0 ? "none" : string.Join(", ", unaccounted));

        Output.Note(
            "Checking only the HTTP status would read this as a clean pass over three\n"
            + "organizations. It is a clean pass over two, and a miss on the third.");
    }
}

File: java/examples/src/main/java/org/pactman/nonprofitcheckplus/examples/cases/Ex19BulkPartialSuccess.java
package org.pactman.nonprofitcheckplus.examples.cases;

import java.util.Arrays;
import java.util.List;
import org.pactman.nonprofitcheckplus.devtools.FixtureEins;
import org.pactman.nonprofitcheckplus.examples.support.Example;
import org.pactman.nonprofitcheckplus.examples.support.ExampleContext;
import org.pactman.nonprofitcheckplus.examples.support.ExampleFailedException;
import org.pactman.nonprofitcheckplus.examples.support.Output;
import org.pactman.nonprofitcheckplus.models.ApiErrorDetail;
import org.pactman.nonprofitcheckplus.models.BulkCheckResult;
import org.pactman.nonprofitcheckplus.models.Nonprofit;

/**
 * EX-19 — Partial success in a batch.
 *
 * <p>A batch where some EINs matched is a successful response with the misses
 * listed inside it — not an exception. Code that only looks at the happy path
 * will process the matches and never notice the rest.
 */
public final class Ex19BulkPartialSuccess implements Example {

    @Override
    public String id() {
        return "ex-19";
    }

    @Override
    public String title() {
        return "A 200 that still reports failures";
    }

    @Override
    public void run(String[] args) {
        List<String> eins = Arrays.asList(
                FixtureEins.PUBLIC_CHARITY, FixtureEins.NO_RECORD, FixtureEins.PUBLIC_CHARITY_SECOND);

        try (ExampleContext context = ExampleContext.withFixtures()) {
            BulkCheckResult result = context.client().nonprofits().checkBulk(eins);

            Output.heading("The response");
            Output.field("HTTP status", result.getStatus());
            Output.field("organizations", result.getOrganizations().size());
            Output.field("notFoundEins", result.getNotFoundEins());
            Output.field("errors", result.getErrors().size());
            Output.field("checkCount", result.getCheckCount());

            if (result.getStatus() != 200) {
                throw new ExampleFailedException("Partial success should be an HTTP 200.");
            }

            if (result.getNotFoundEins().isEmpty()) {
                throw new ExampleFailedException("The missing EIN should have been reported.");
            }

            Output.heading("What the errors array carried");

            for (ApiErrorDetail detail : result.getErrors()) {
                Output.field("resource", detail.getResource());
                Output.field("code", detail.getCode());
                Output.field("reason", detail.getReason());
                Output.field("eins", detail.getEins());
            }

            Output.heading("Matched");

            for (Nonprofit organization : result.getOrganizations()) {
                Output.bullet(organization.getEin() + " — " + organization.getOrganizationName());
            }
        }

        Output.note("Three EINs, two records, one 200. getNotFoundEins() is collected from the\n"
                + "envelope's errors array, so the misses are as visible as the matches.\n\n"
                + "Only matched EINs are billed, which is why checkCount rose by two and not by\n"
                + "three — and why usage cannot be reconstructed from the size of the batch you\n"
                + "sent.");
    }
}
